### PR TITLE
[compiler-rt][nsan] Add support for nan detection

### PR DESCRIPTION
**Pull request by cseslowpoke**
Add support for nan detection
#100305

### CODE REVIEW

**Comment by cseslowpoke** [comment]
@alexander-shaposhnikov

**Comment by alexander-shaposhnikov** [comment]
Have not looked at the code yet.
Regarding tests - could you please add larger tests (e.g. quadratic equation and unstable version of softmax).
P.S. It'd be interesting to try it on a real-world program and see what the reports look like / check if things work as expected or not.

File: compiler-rt/lib/nsan/nsan_flags.inc [hunk 1]
```diff
@@ -47,3 +47,5 @@ NSAN_FLAG(bool, enable_loadtracking_stats, false,
           "memory, the number of times nsan resumed from the original value "
           "due to invalid or unknown types.")
 NSAN_FLAG(bool, print_stats_on_exit, false, "If true, print stats on exit.")
+NSAN_FLAG(bool, check_nan, false,
```
Comment (chestnykh) [inline compiler-rt/lib/nsan/nsan_flags.inc:50]: Probably true by default, i dont think anyone want see nans in the program

File: compiler-rt/test/nsan/nan.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clangxx_nsan -O0 -g %s -o %t
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/nan.cpp:1]: need to make this option work with `NSAN_OPTIONS=halt_on_error=0` and `NSAN_OPTIONS=halt_on_error=1`
(the behavior should be similar to the current consistency check)
(and cover by tests)

**Comment by cseslowpoke** [comment]
> Have not looked at the code yet. Regarding tests - could you please add larger tests (e.g. quadratic equation and unstable version of softmax). P.S. It'd be interesting to try it on a real-world program and see what the reports look like / check if things work as expected or not.
Sorry for taking so long to submit this. In the past few days, I wrote some tests (maybe I can add a few more) and made it so that the stack is output when NaN is detected.

File: compiler-rt/test/nsan/softmax.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=stable_softmax %s -o %t
+// RUN: NSAN_OPTIONS=halt_on_error=1,log2_max_relative_error=19 %run %t 2>&1
+
+#include<iostream>
+#include<vector>
+#include<algorithm>
+#include<cmath>
+// unstable softmax
+template <typename T>
+__attribute__((noinline)) void softmax(std::vector<T> &values) {
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/softmax.cpp:16]: const std::vector<T> &values

File: compiler-rt/test/nsan/softmax.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=stable_softmax %s -o %t
+// RUN: NSAN_OPTIONS=halt_on_error=1,log2_max_relative_error=19 %run %t 2>&1 
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=stable_softmax %s -o %t
+// RUN: NSAN_OPTIONS=halt_on_error=1,log2_max_relative_error=19 %run %t 2>&1
+
+#include<iostream>
+#include<vector>
+#include<algorithm>
+#include<cmath>
+// unstable softmax
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/softmax.cpp:14]: nit: blank line after includes

File: compiler-rt/test/nsan/nan.cpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: NSAN_OPTIONS=check_nan=true %run %t 2>&1 | FileCheck %s
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/nan.cpp:2]: Could you please also test halt_on_error=0 / halt_on_error=1 here ? (similarly to the other test)

**Comment by alexander-shaposhnikov** [comment]
@cseslowpoke, thanks for working on this. I've added a few comments,
I'd probably wait for a few days to let @MaskRay have a look at the patch as well.

File: compiler-rt/lib/nsan/nsan_flags.inc [hunk 6]
```diff
@@ -47,3 +47,5 @@ NSAN_FLAG(bool, enable_loadtracking_stats, false,
           "memory, the number of times nsan resumed from the original value "
           "due to invalid or unknown types.")
 NSAN_FLAG(bool, print_stats_on_exit, false, "If true, print stats on exit.")
+NSAN_FLAG(bool, check_nan, true,
```
Comment (alexander-shaposhnikov) [inline compiler-rt/lib/nsan/nsan_flags.inc:50]: i remember the comment about making `check_nan` `true`,  I'm a bit worried that this check might fire early in places where nans are legitimate... Not sure, but intuitively I'd probably keep it "false" by default, at least for now.

Comment (cseslowpoke) [reply compiler-rt/lib/nsan/nsan_flags.inc:50]: Sure, I will change this option.

**Comment by alexander-shaposhnikov** [comment]
@cseslowpoke - would be good to add a test where nans are produced by vector instructions

**Comment by cseslowpoke** [comment]
> @cseslowpoke - would be good to add a test where nans are produced by vector instructions

How should I generate vector instructions? Should I use SIMD instructions, or is there another approach I should take?

**Comment by alexander-shaposhnikov** [comment]
I've looked at the tests for `msan`, there are a few that use SSE, I'd add a small test along those lines. Adding a larger test
(e.g. Gaussian elimination) is more than welcome too. 
Thanks for your patience.

File: compiler-rt/test/nsan/softmax.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=softmax %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0,log2_max_relative_error=19 %run %t 2>&1 | FileCheck %s
+
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=softmax %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0,log2_max_relative_error=19 %run %t 2>&1 | FileCheck %s
+
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=stable_softmax %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0,log2_max_relative_error=19 %run %t 2>&1 
+
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=stable_softmax %s -o %t
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/softmax.cpp:10]: i suspect the second invocation was meant for `halt_on_error=1`.
In this case we don't need to rebuild the binary, and if the exit code is nonzero you may want to use `not` (take a look at similar tests)

Comment (cseslowpoke) [reply compiler-rt/test/nsan/softmax.cpp:10]: Okay, I'll change it to that.

File: compiler-rt/test/nsan/softmax.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=softmax %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0,log2_max_relative_error=19 %run %t 2>&1 | FileCheck %s
+
+// RUN: %clangxx_nsan -O0 -g -DSOFTMAX=softmax %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0,log2_max_relative_error=19 %run %t 2>&1 | FileCheck %s
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/softmax.cpp:5]: this is a duplicate of the previous `run`

**Comment by cseslowpoke** [comment]
I tried a simple example, but when I use -mavx and Clang vector extension to generate SIMD instructions along with nsan, a segmentation fault occurs.
```cpp
#include <iostream>
#include <cmath>

typedef float v8sf __attribute__ ((vector_size(32)));


v8sf simd_sqrt(v8sf a) {
  return __builtin_elementwise_sqrt(a);
}

int main() {
  v8sf a = {1.0, -2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0};
  a = simd_sqrt(a);
  return 0;
}
```

**Comment by alexander-shaposhnikov** [comment]
Interesting, would be good to investigate what's going on there.
If you get stuck - let me know,  I'll take a look.

**Comment by cseslowpoke** [comment]
I tried adding `#include <immintrin.h>` to use SIMD instructions, but I'm not sure where the warning should appear.

File: compiler-rt/test/nsan/vec_sqrt.cpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+#include <immintrin.h>
+#include <iostream>
+
+void simd_sqrt(const float *input, float *output, size_t size) {
+  size_t i = 0;
+  for (; i + 7 < size; i += 8) {
+    __m256 vec = _mm256_loadu_ps(&input[i]);
+    __m256 result = _mm256_sqrt_ps(vec);
+    _mm256_storeu_ps(&output[i], result);
+    // when simd is used, the warning is not triggered
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/vec_sqrt.cpp:16]: it looks like something is broken.
we'd like to detect cases like this, the report below  triggered by std::sqrt below is not relevant, need to figure out/understand what's broken for simd.

Comment (cseslowpoke) [reply compiler-rt/test/nsan/vec_sqrt.cpp:16]: Sure, I’ll check the generated LLVM IR to identify where the issue might be.

File: compiler-rt/test/nsan/nan.cpp (hunk 2 above)
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/nan.cpp:1]: We also need to add a run for -O3.
See https://stackoverflow.com/questions/54481855/clang-ignoring-attribute-noinline if optimizations trigger constant folding here.
I'd recommend a slight modification (probably should suffice to make things work):
```
__attribute__((noinline)) float f(float p, float q) {
     float  ret = p / q;
     return ret;
}
```
And then call f(0., 0.)

**Comment by alexander-shaposhnikov** [comment]
@cseslowpoke 
I've applied locally your patch on top the current trunk and tried the example
```
#include <iostream>
#include <cmath>

typedef float v8sf __attribute__ ((vector_size(32)));

v8sf simd_sqrt(v8sf a) {
  return __builtin_elementwise_sqrt(a);
}

int main() {
  v8sf a = {1.0, -2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0};
  a = simd_sqrt(a);

  // This prevents DCE.
  for (size_t i = 0; i < 8; ++i) {
    std::cout << a[i] << std::endl;
  }

  return 0;
}

```

```
clang++ -fsanitize=numerical -mavx -O3 -g ~/simd.cpp -o ~/simd.exe
NSAN_OPTIONS=check_nan=true,halt_on_error=0 ~/simd.exe
```
works as expected.
Let's use it as a test for now.

File: compiler-rt/lib/nsan/nsan.cpp [hunk 10]
```diff
@@ -444,6 +444,25 @@ int32_t checkFT(const FT value, ShadowFT Shadow, CheckTypeT CheckType,
...
+      return flags().resume_after_suppression ? kResumeFromValue
+                                              : kContinueWithShadow;
+    }
+    Decorator D;
+    Printf("%s", D.Warning());
+    Printf("WARNING: NumericalStabilitySanitizer: NaN detected\n");
+    Printf("%s", D.Default());
+    stack.Print();
+    // Performing other tests for NaN values is meaningless when dealing with numbers.
+    return kResumeFromValue;
```
Comment (alexander-shaposhnikov) [inline compiler-rt/lib/nsan/nsan.cpp:463]: this doesn't respect the option `halt_on_error`, see
```
 if (flags().halt_on_error) {
    if (common_flags()->abort_on_error)
      Printf("ABORTING\n");
    else
      Printf("Exiting\n");
    Die();
  }
```
below (lines 575 - 595).
We need the same here + need to test it.

File: compiler-rt/test/nsan/vec_sqrt.cpp (hunk 9 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/test/nsan/vec_sqrt.cpp:16]: I've simplified the test a bit:
```
#include <cassert>
#include <cmath>
#include <immintrin.h>
#include <iostream>

void simd_sqrt(const float *input, float *output, size_t size) {
  size_t i = 0;
  assert(size % 8 == 0);
  for (; i < size; i += 8) {
    __m256 vec = _mm256_loadu_ps(&input[i]);
    __m256 result = _mm256_sqrt_ps(vec);
    _mm256_storeu_ps(&output[i], result);
  }
}

int main() {
  float input[] = {0.0, 1.0, -2.0, 3.0, 4.0, 5.0, 6.0, 7.0};
  float output[8];

  simd_sqrt(input, output, 8);

  for (int i = 0; i < 8; ++i) {
    std::cout << output[i] << std::endl;
  }
  return 0;
}
```

```
clang++ -fsanitize=numerical -mavx -O3 -g ~/simd2.cpp -o ~/simd2.exe
```

It also works as expected.

**Comment by alexander-shaposhnikov** [comment]
@cseslowpoke, let's try to complete this effort:
0. Make sure halt_on_error works (see comments above)
1. Update tests (see inline comments)
2. Test halt_on_erro=1 and test -O3

File: compiler-rt/test/nsan/vec_sqrt.cpp (hunk 9 above)
Comment (cseslowpoke) [reply compiler-rt/test/nsan/vec_sqrt.cpp:16]: When using `sqrt`, it detects `nan`, but I'm not sure if using `_mm256_sqrt_ps` would exhibit similar behavior.

File: compiler-rt/test/nsan/nan.cpp [hunk 11]
```diff
@@ -1,19 +1,22 @@
 // RUN: %clangxx_nsan -O0 -g %s -o %t
 // RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0 %run %t 2>&1 | FileCheck %s
 
+// RUN: %clangxx_nsan -O3 -g %s -o %t
+// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=0 %run %t 2>&1 | FileCheck %s
+
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/nan.cpp:6]: let's add 
```
// RUN: NSAN_OPTIONS=check_nan=true,halt_on_error=1 not %run %t
```

File: compiler-rt/test/nsan/clang_ext.cpp [hunk 12]
```diff
@@ -0,0 +1,25 @@
+// RUN: %clangxx_nsan -O0 -g -mavx %s -o %t
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/clang_ext.cpp:1]: nit: rename this test to smth like vec_sqrt_ext.cpp

**Review by alexander-shaposhnikov — APPROVED** [review]
Added a few comments, 
otherwise, LGTM.
We can merge it in,
I'll probably do some cleanup afterwards.

File: compiler-rt/test/nsan/softmax.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    for (auto &i: values) {
+      i /= sum_exp;
+    }
+}
+
+// use max value to avoid overflow
+// \sigma_i exp(x_i) / \sum_j exp(x_j) = \sigma_i exp(x_i - max(x)) / \sum_j exp(x_j - max(x))
+template <typename T>
+__attribute__((noinline)) void stable_softmax(std::vector<T> values) {
```
Comment (alexander-shaposhnikov) [inline compiler-rt/test/nsan/softmax.cpp:34]: missing ref: &values

**Comment by MatzeB** [comment]
Hi, I believe this change (after re-applied in 51365212362c4d0e32a0c747ab85bbf3919944b8 ) fails our builds using gcc 11.x:

```
/home/engshare/third-party2/gcc/11.x/centos9-native/886b5eb/bin/g++ -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/engshare/third-party2/llvm-fb/17/src/build-platform010/pic/projects/compiler-rt/lib/nsan -I/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan -I/home/engshare/third-party2/llvm-fb/17/src/build-platform010/pic/include -I/home/engshare/third-party2/llvm-fb/17/src/llvm-project/llvm/include -I/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/.. -O3 -g -pipe -Wall -Wno-error -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fno-omit-frame-pointer -momit-leaf-frame-pointer -fno-reorder-blocks-and-partition -gz=none -Wl,-z,relro -fcommon -gdwarf-4 -gstrict-dwarf -march=corei7 -mtune=skylake -nostdinc -grecord-gcc-switches -fdebug-prefix-map=/home/engshare/third-party2/llvm-fb/17/src/llvm-project=/home/engshare/third-party2/llvm-fb/17/src/llvm-project -fdebug-prefix-map=/home/engshare/third-party2/llvm-fb/17/src/build-platform010/pic=/home/engshare/third-party2/llvm-fb/17/src/llvm-project -isystem/home/engshare/third-party2/libedit/3.1/platform010/6e03f73/include -isystem/home/engshare/third-party2/zlib/1.2.8/platform010/76ebdda/include -isystem/home/engshare/third-party2/ncurses/6.1/platform010/76ebdda/include -isystem/home/engshare/third-party2/xz/5.2.2/platform010/76ebdda/include -isystem/home/engshare/third-party2/readline/8.0/platform010/6e03f73/include -isystem/home/engshare/third-party2/gdbm/1.11/platform010/76ebdda/include -isystem/home/engshare/third-party2/binutils/2.37/platform010/64091f4/include -isystem/home/engshare/third-party2/openssl/1.1.1/platform010/76ebdda/include -isystem/home/engshare/third-party2/sqlite/3.36/platform010/9079c97/include -isystem/home/engshare/third-party2/python/3.10/platform010/9168809/include -isystem/home/engshare/third-party2/libffi/3.2.1/platform010/76ebdda/lib/libffi-3.2.1/include -isystem/home/engshare/third-party2/jemalloc/master/platform010/779a252/include -isystem/home/engshare/third-party2/libuuid/1.0.2/platform010/76ebdda/include -isystem/home/engshare/third-party2/libxml2/2.9.9/platform010/ec4b15a/include -isystem/home/engshare/third-party2/libxml2/2.9.9/platform010/ec4b15a/include/libxml2 -isystem/home/engshare/third-party2/grpc/1.42.0/platform010/92e86c7/include -isystem/home/engshare/third-party2/bzip2/1.0.6/platform010/76ebdda/include -isystem/home/engshare/third-party2/libunwind/1.8/platform010/76ebdda/include -isystem/home/engshare/third-party2/libzip/1.5.1/platform010/6ace573/include -isystem/home/engshare/third-party2/benchmark/trunk/platform010/76ebdda/include -isystem/home/engshare/third-party2/re2/20190601/platform010/76ebdda/include -isystem/home/engshare/third-party2/gflags/2.2.0/platform010/76ebdda/include -isystem/home/engshare/third-party2/c-ares/1.13.0/platform010/76ebdda/include -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O2 -g -DNDEBUG -std=c++17 -fvisibility=default -m64 -fPIC -fno-builtin -fno-exceptions -fomit-frame-pointer -funwind-tables -fno-stack-protector -fvisibility=hidden -fno-lto -O3 -g -Wno-variadic-macros -nostdinc++ -fno-rtti -MD -MT projects/compiler-rt/lib/nsan/CMakeFiles/RTNsan.x86_64.dir/nsan.cpp.o -MF projects/compiler-rt/lib/nsan/CMakeFiles/RTNsan.x86_64.dir/nsan.cpp.o.d -o projects/compiler-rt/lib/nsan/CMakeFiles/RTNsan.x86_64.dir/nsan.cpp.o -c /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp
In file included from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_linux.h:22,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_procmaps.h:25,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_allocator.h:24,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_allocator_internal.h:16,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_addrhashmap.h:19,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan_stats.h:19,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp:35:
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/../sanitizer_common/sanitizer_platform_limits_posix.h:593:10: warning: ISO C++ prohibits anonymous structs [-Wpedantic]
  593 |   struct {
      |          ^
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp: In function 'void __nsan_dump_shadow_mem(const u8*, size_t, size_t, size_t)':
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp:359:23: warning: cast from type 'const u8*' {aka 'const unsigned char*'} to type 'void*' casts away qualifiers [-Wcast-qual]
  359 |     printf("%p:    ", (void *)(addr + R * bytes_per_line));
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp: In instantiation of 'int32_t checkFT(FT, ShadowFT, __nsan::CheckTypeT, __sanitizer::uptr) [with FT = double; ShadowFT = __float128; int32_t = int; __sanitizer::uptr = long unsigned int]':
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp:605:17:   required from here
/home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp:449:33: error: call of overloaded 'isnan(const InternalFT&)' is ambiguous
  449 |   if (flags().check_nan && isnan(check_value)) {
      |                            ~~~~~^~~~~~~~~~~~~
In file included from /home/engshare/third-party2/libgcc/11.x/platform010/5684a5a/include/c++/11.x/math.h:36,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.h:29,
                 from /home/engshare/third-party2/llvm-fb/17/src/llvm-project/compiler-rt/lib/nsan/nsan.cpp:33:
/home/engshare/third-party2/libgcc/11.x/platform010/5684a5a/include/c++/11.x/cmath:624:3: note: candidate: 'constexpr bool std::isnan(long double)'
  624 |   isnan(long double __x)
      |   ^~~~~
/home/engshare/third-party2/libgcc/11.x/platform010/5684a5a/include/c++/11.x/cmath:619:3: note: candidate: 'constexpr bool std::isnan(double)'
  619 |   isnan(double __x)
      |   ^~~~~
/home/engshare/third-party2/libgcc/11.x/platform010/5684a5a/include/c++/11.x/cmath:611:3: note: candidate: 'constexpr bool std::isnan(float)'
  611 |   isnan(float __x)
      |   ^~~~~
```

could you take a look?

**Comment by alexander-shaposhnikov** [comment]
@MatzeB - thanks, that's interesting, if you have a minute, could you check if `::isnan` would build successfully ?

**Comment by cseslowpoke** [comment]
I think it's because `std::isnan` doesn't provide a version for `__float128`, so we may need to find an alternative way to check for `nan`.

**Comment by alexander-shaposhnikov** [comment]
I don't have a local repro, but based on the provided log - `FT = double`, and the current isnan is actually a macro
that expands to `__builtin_isnan`. I'm not sure why this doesn't compile, but I'd try the ::isnan. If this doesn't work, one can use smth like `bool isNaN(T t) { return t != t; }`

**Pull request merged**